modules/gapi/src/backends/oak/goakbackend.cpp

```diff
@@ -0,0 +1,691 @@
+                }
+                curr_arg_idx += op.args.size();
+            }
+        }
+
+        // Check that all inputs and outputs are properly filled after constructing kernels
+        // to then link it together
+        for (const auto& node : m_oak_nodes) {
+            auto el = node.second;
+            if (std::any_of(el.inputs.cbegin(), el.inputs.cend(),
+                            [](ExtractTypeHelper::InputPtr ptr) {
+                    return ptr == nullptr;
+                })) {
+                GAPI_Assert(false && "DAI input are not set");
+            }
+            if (std::any_of(el.outputs.cbegin(), el.outputs.cend(),
+                            [](ExtractTypeHelper::OutputPtr ptr) {
+                    return ptr == nullptr;
+                })) {
+                GAPI_Assert(false && "DAI outputs are not set");
+            }
+        }
+
+        // Properly link all nodes
+        // 1. Link input nodes to camera
+        for (const auto& d : ins_data)
+        {
+            for (const auto& nh : nodes)
+            {
+                if (m_gm.metadata(nh).contains<cv::gimpl::Op>())
+                {
+                    for (const auto& indata : nh.get()->inNodes()) {
+                        auto rc = m_gm.metadata(indata).get<cv::gimpl::Data>().rc;
+                        if (rc == d.rc)
+                        {
+                            GAPI_Assert(m_oak_nodes.find(nh) != m_oak_nodes.end());
+                            GAPI_Assert(m_oak_nodes[nh].inputs.size() == 1);
+                            // FIXME: covert other camera outputs
+                            m_camera_input->video.link(*(m_oak_nodes[nh].inputs[0]));
+                            m_oak_nodes[nh].processed = true;
+                        }
+                    }
+                }
+            }
+        }
+
+        // 2. Link output nodes to XLinkOut nodes
+        size_t out_counter = 0;
+        for (const auto& d : outs_data)
+        {
+            for (const auto& nh : nodes)
+            {
+                if (m_gm.metadata(nh).contains<cv::gimpl::Op>())
+                {
+                    for (const auto& outdata : nh.get()->outNodes()) {
+                        auto rc = m_gm.metadata(outdata).get<cv::gimpl::Data>().rc;
+                        if (rc == d.rc)
+                        {
+                            GAPI_Assert(m_oak_nodes.find(nh) != m_oak_nodes.end());
+                            GAPI_Assert(m_oak_nodes[nh].outputs.size() == 1);
+                            GAPI_Assert(out_counter < m_xlink_outputs.size());
+                            m_oak_nodes[nh].outputs[0]->link(m_xlink_outputs[out_counter++]->input);
+
+                            if (!m_oak_nodes[nh].processed) {
+                                LinkToParentHelper(nh, nodes);
+                                m_oak_nodes[nh].processed = true;
+                            }
```

Comment:
This `if` can be removed since the same is done in the loop below

Reply:
Reworked